downloader/src/main/kotlin/vcs/GitRepo.kt

```diff
@@ -89,6 +90,66 @@ class GitRepo : GitBase() {
 
     override fun isApplicableUrlInternal(vcsUrl: String) = false
 
+    override fun initWorkingTree(targetDir: File, vcs: VcsInfo): WorkingTree {
+        val manifestRevision = vcs.revision.takeUnless { it.isBlank() } ?: "master"
+        val manifestPath = vcs.path.takeUnless { it.isBlank() } ?: "default.xml"
+
+        log.info {
+            "Initializing git-repo from ${vcs.url} with revision '$manifestRevision' and manifest '$manifestPath'."
+        }
+
+        // Clone all projects instead of only those in the "default" group until we support specifying groups.
+        runRepoCommand(
+            targetDir,
+            "init", "--groups=all", "--no-repo-verify",
+            "--no-clone-bundle", "--repo-branch=$GIT_REPO_BRANCH",
+            "-b", manifestRevision,
+            "-u", vcs.url,
+            "-m", manifestPath
+        )
+
+        return getWorkingTree(targetDir)
+    }
+
+    override fun updateWorkingTree(
+        workingTree: WorkingTree,
+        revision: String,
+        path: String,
+        recursive: Boolean
+    ): Boolean {
+        val manifestRevision = revision.takeUnless { it.isBlank() } ?: "master"
+        val manifestPath = path.takeUnless { it.isBlank() } ?: "default.xml"
+
+        return try {
+            // Switching manifest branches / revisions requires running "init" again.
+            runRepoCommand(workingTree.workingDir, "init", "-b", manifestRevision, "-m", manifestPath)
+
+            // Repo allows to checkout Git repositories to nested directories. If a manifest is badly configured, a
+            // nested Git checkout overwrites files in a directory of the upper-level Git repository. However, we still
+            // want to be able to download such projects, so specify "--force-sync" to work around that issue.
+            val syncArgs = mutableListOf("sync", "-c", "--force-sync")
+
+            if (recursive) {
+                syncArgs += "--fetch-submodules"
+            }
+
+            runRepoCommand(workingTree.workingDir, *syncArgs.toTypedArray())
+
+            log.debug { runRepoCommand(workingTree.workingDir, "info").stdout }
+
+            true
+        } catch (e: IOException) {
+            e.showStackTrace()
+
+            log.warn {
+                "Failed to sync the working tree to revision '$manifestRevision' using manifest '$manifestPath': " +
+                        e.message
```

Comment:
Should be `e.collectMessagesAsString()`.

Reply:
Done.